### PR TITLE
Patch 3 for program from program creation

### PR DESCRIPTION
**Pull request by SabaunT**
Creating program from program implementation is on st-create-program-from-program-iter1 branch and quite big. To make it easy for review, I decided to divide it to separate PRs.

This is the third one.

**Release Notes**: `the patch defines some logic to test program from program creation in `gear-test`. An example for `create_program_with_gas` was also introduced. 

@gear-tech/dev

### CODE REVIEW

File: examples/create-program/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            debug!("Sent to a new program message with id {:?}", msg_id);
+        }
+        _ => {
+            panic!("Unknown option");
+        }
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {}
```
Comment (shamilsan) [inline examples/create-program/src/lib.rs:73]: Empty `init` again?
```suggestion
```

**Review by breathx — APPROVED** [review]

File: gear-test/spec/test_create_program.yaml [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+          - destination: 
+              kind: ss58
+              value: 5Fypuf8SCGzqtJy91VJqb17WgUbgVTC6k3Vb4yfj3VBj2at1
+            init: true
+            payload:
+              kind: utf-8
+              value: unique
+          - destination: 
+              kind: ss58
+              value: 5Fypuf8SCGzqtJy91VJqb17WgUbgVTC6k3Vb4yfj3VBj2at1
```
Comment (shamilsan) [inline gear-test/spec/test_create_program.yaml:73]: We have anchors in YAML for such kind repeating objects, please take it into account in the future.
```suggestion
          - destination: &address
              kind: ss58
              value: 5Fypuf8SCGzqtJy91VJqb17WgUbgVTC6k3Vb4yfj3VBj2at1
```

File: gear-test/spec/test_create_program.yaml [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        programs:
+          - kind: id
+            value: 1
+          - kind: ss58
+            # computed manually new program address
+            value: 5Fypuf8SCGzqtJy91VJqb17WgUbgVTC6k3Vb4yfj3VBj2at1
+        messages:
+          - destination: 
+              kind: ss58
+              value: 5Fypuf8SCGzqtJy91VJqb17WgUbgVTC6k3Vb4yfj3VBj2at1
```
Comment (shamilsan) [inline gear-test/spec/test_create_program.yaml:86]:
```suggestion
          - destination: *address
```

File: gear-test/src/manager.rs [hunk 4]
```diff
@@ -187,27 +203,39 @@ impl JournalHandler for InMemoryExtManager {
...
-        &mut self,
-        _code_hash: gear_core::program::CodeHash,
-        _candidates: Vec<(ProgramId, MessageId)>,
-    ) {
-        todo!() // Patch 3 is coming
+    fn store_new_programs(&mut self, code_hash: CodeHash, candidates: Vec<(ProgramId, MessageId)>) {
+        if let Some(code) = self.codes.get(&code_hash).cloned() {
+            for (candidate_id, init_message_id) in candidates {
+                if !self.actors.contains_key(&candidate_id) {
+                    let program = Program::new(candidate_id, code.clone())
```
Comment (shamilsan) [inline gear-test/src/manager.rs:224]: Do we have two clones for `code` here, isn't it?

**Review by shamilsan — APPROVED** [review]
Great job! 👍 Don't forget to add release notes for it.

File: gear-test/src/manager.rs (hunk 4 above)
Comment (SabaunT) [reply gear-test/src/manager.rs:224]: Nope. Without `cloned` we can't mutate `self` in 226.

File: gear-test/spec/test_create_program.yaml (hunk 3 above)
Comment (SabaunT) [reply gear-test/spec/test_create_program.yaml:84]: done

File: gear-test/spec/test_create_program.yaml (hunk 2 above)
Comment (SabaunT) [reply gear-test/spec/test_create_program.yaml]: done

**Pull request merged**